java/core/src/main/java/com/microsoft/bond/BondType.java

```diff
@@ -0,0 +1,514 @@
+        sb.append(this.getQualifiedName());
+        BondType<?>[] typeArguments = this.getGenericTypeArguments();
+        if (typeArguments != null) {
+            sb.append("<");
+            sb.append(typeArguments[0].getFullName());
+            for (int i = 1; i < typeArguments.length; ++i) {
+                sb.append(", ");
+                sb.append(typeArguments[i].getFullName());
+            }
+            sb.append(">");
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Gets the {@link BondDataType} value for this type.
+     *
+     * @return the bond data type enumeration value
+     */
+    public abstract BondDataType getBondDataType();
+
+    /**
+     * Gets the {@link Class} instance for the values (objects) of this type.
+     * The actual class of a value of this Bond type may not be the same as the returned class,
+     * but its is always assignable to it (i.e. a subclass or a Java primitive type).
+     *
+     * @return the class object
+     */
+    public abstract Class<T> getValueClass();
+
+    /**
+     * Gets the {@link Class} instance for the Java primitive values (non-objects) of this type
+     * or null if a Java primitive type does not exist. This method complements the
+     * {@link BondType#getValueClass()} method which returns the classes for object instances.
+     *
+     * @return the class object for a primitive type or null if a primitive type doesn't exists
+     */
+    public abstract Class<T> getPrimitiveValueClass();
+
+    /**
+     * Gets a value indicating whether values of this type can be assigned to null.
+     *
+     * @return whether value of this type can be assigned to null
+     */
+    public abstract boolean isNullableType();
+
+    /**
+     * Indicates whether this type is a generic type that is specialized with one or more generic type arguments.
+     *
+     * @return true if this type is a specialization of a generic type
+     */
+    public abstract boolean isGenericType();
+
+    /**
+     * Gets a new array instance containing the type descriptors of the generic type arguments or null
+     * if this type is not a specialization of a generic type.
+     *
+     * @return an array containing the descriptors of the generic type arguments or null
+     */
+    public abstract BondType<?>[] getGenericTypeArguments();
+
+    /**
+     * Returns the default value of this type as a shared instance for immutable primitive types
+     * or a new instance for all other types. For non-nullable structs and containers this method
+     * returns an initialized value equivalent to invoking the default constructor. For nullable
+     * values this method returns null.
+     *
+     * @return the default initialized value of this type
+     */
+    protected abstract T newDefaultValue();
+
+    /**
+     * Serializes a value of this type into a protocol writer.
+     * This method is not recommended for Java primitive (non-object) types. To serialize primitive
```

Comment:
There is no "non-object" type; every type has to know how to deal with objects. Certain primitive types (e.g. Int32) also need to know how to deal with primitives. The point of this comment is to avoid unnecessary boxing/unboxing every time.

---
In reply to: [123092975](https://github.com/Microsoft/bond/pull/502#discussion_r123092975) [](ancestors = 123092975)